### ISSUE TITLE
[BUG] ModuleNotFoundError: No module named 'opentelemetry.semconv'

### ISSUE THREAD

**Issue by TomSteenbergen** [mlflow/mlflow#12149]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.13.0
- Tracking server: 2.13.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 14.5
- **Python version**: 3.10
- **yarn version, if running the dev UI**:

### Describe the problem

When importing `mlflow`, the following error is raised: `ModuleNotFoundError: No module named 'opentelemetry.semconv'`

This is due to `mlflow` using `opentelemetry` and `opentelemetry` moved the corresponding `semconv` module in version `1.24` to a different place, resulting in this `ModuleNotFoundError`.

We should probably update the `opentelemetry-sdk` dependency constraint to `<1.24` to prevent downstream users of running into this error.

### Code to reproduce issue

```
import mlflow
```

### Stack trace

```
ModuleNotFoundError: No module named 'opentelemetry.semconv'
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
cc @B-Step62

**Comment by harupy** [comment]
I was able to reproduce the same error after _**manually**_ uninstalling `opentelemetry-semantic-conventions`:

```
root@1a5ab59a3a4b:/# pip uninstall opentelemetry-semantic-conventions
Found existing installation: opentelemetry-semantic-conventions 0.45b0
Uninstalling opentelemetry-semantic-conventions-0.45b0:
  Would remove:
    /usr/local/lib/python3.9/site-packages/opentelemetry/semconv/*
    /usr/local/lib/python3.9/site-packages/opentelemetry_semantic_conventions-0.45b0.dist-info/*
Proceed (Y/n)? y
  Successfully uninstalled opentelemetry-semantic-conventions-0.45b0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@1a5ab59a3a4b:/# python -c "import mlflow"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/mlflow/__init__.py", line 34, in <module>
    from mlflow import (
  File "/usr/local/lib/python3.9/site-packages/mlflow/artifacts/__init__.py", line 11, in <module>
    from mlflow.tracking import _get_store
  File "/usr/local/lib/python3.9/site-packages/mlflow/tracking/__init__.py", line 8, in <module>
    from mlflow.tracking._model_registry.utils import (
  File "/usr/local/lib/python3.9/site-packages/mlflow/tracking/_model_registry/utils.py", line 4, in <module>
    from mlflow.store.db.db_types import DATABASE_ENGINES
  File "/usr/local/lib/python3.9/site-packages/mlflow/store/__init__.py", line 1, in <module>
    from mlflow.store import _unity_catalog  # noqa: F401
  File "/usr/local/lib/python3.9/site-packages/mlflow/store/_unity_catalog/__init__.py", line 1, in <module>
    from mlflow.store._unity_catalog import registry  # noqa: F401
  File "/usr/local/lib/python3.9/site-packages/mlflow/store/_unity_catalog/registry/__init__.py", line 1, in <module>
    from mlflow.store._unity_catalog.registry import rest_store  # noqa: F401
  File "/usr/local/lib/python3.9/site-packages/mlflow/store/_unity_catalog/registry/rest_store.py", line 9, in <module>
    from mlflow.entities import Run
  File "/usr/local/lib/python3.9/site-packages/mlflow/entities/__init__.py", line 23, in <module>
    from mlflow.entities.span import LiveSpan, NoOpSpan, Span, SpanType
  File "/usr/local/lib/python3.9/site-packages/mlflow/entities/span.py", line 7, in <module>
    from opentelemetry.sdk.trace import Event as OTelEvent
  File "/usr/local/lib/python3.9/site-packages/opentelemetry/sdk/trace/__init__.py", line 59, in <module>
    from opentelemetry.sdk.resources import Resource
  File "/usr/local/lib/python3.9/site-packages/opentelemetry/sdk/resources/__init__.py", line 74, in <module>
    from opentelemetry.semconv.resource import ResourceAttributes
ModuleNotFoundError: No module named 'opentelemetry.semconv'
```

**Comment by harupy** [comment]
@TomSteenbergen any updates on this?

**Comment by harupy** [comment]
No reply from the author, closing.

**Comment by TomSteenbergen** [comment]
Hi @harupy ! Excuse me for the late reply, this issue thread fell through the cracks.

I tried to replicate the issue, but unfortunately I wasn't able to. I do notice from this [link](https://github.com/mlflow/mlflow/issues/12149#issuecomment-2135303984) that the `opentelemetry-semantic-conventions` constraint in `opentelemetry-sdk` is set to `0.45b0.dev`, which was released on the day I filed this issue. Perhaps it was a simple timing issue.

Fine to close this issue for now!
